src/Http/Middleware/ProxyAddress.php

```diff
@@ -0,0 +1,59 @@
+<?php
+
+/*
+ * This file is part of Flarum.
+ *
+ * For detailed copyright and license information, please view the
+ * LICENSE file that was distributed with this source code.
+ */
+
+namespace Flarum\Http\Middleware;
+
+use Flarum\Http\Exception\ProxyNotAllowedException;
+use Illuminate\Support\Arr;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface as Middleware;
+use Psr\Http\Server\RequestHandlerInterface;
+
+class ProxyAddress implements Middleware
+{
+    /**
+     * @var bool
+     */
+    protected $enabled;
+
+    /**
+     * @var array
+     */
+    protected $allowedAddresses;
+
+    /**
+     * @param bool $enabled
+     * @param array $allowedAddresses
+     */
+    public function __construct($enabled, $allowedAddresses)
+    {
+        $this->enabled = $enabled;
+        $this->allowedAddresses = $allowedAddresses;
+    }
+
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
+    {
+        $ipAddress = Arr::get($request->getServerParams(), 'REMOTE_ADDR', '127.0.0.1');
+
+        if ($this->enabled) {
+            if (in_array($ipAddress, $this->allowedAddresses)) {
+                // standard header for proxies, see: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For
+                $ipAddress = Arr::get($request->getServerParams(), 'X_FORWARDED_FOR', $ipAddress);
+                $ipAddress = Arr::get($request->getServerParams(), 'HTTP_CLIENT_IP', $ipAddress);
```

Comment:
Is there a reason why these are in the order they are in?

Reply:
I've used the order "cave bro" used in his basic implementation.

Reply:
Also found a more appropriate source: https://serverfault.com/a/920060

Reply:
Might it also be worth adding X-ProxyUser-Ip as per the linked documentation (https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For)? Also (asking as I haven't dealt with this myself), do you know of any big proxies that might set multiple of these headers?

Reply:
While I agree that the "Forwarded" header is the new standard and we should probably support it, I have a few concerns:
1. The header has a key-value map which would we need to parse. Maybe PHP has that already built-in?
2. I doubt that proxies will ever stop sending headers that were used for this purpose before, just because of support for older applications.

Reply:
Yeah we sure can add this header.